solarforecastarbiter/datamodel.py

```diff
@@ -419,6 +419,130 @@ class Observation(BaseModel):
     __post_init__ = __set_units__
 
 
+@dataclass(frozen=True)
+class AggregateObservation(BaseModel):
+    """
+    Class for keeping track of Observations in Aggregates
+    and when they are added and removed from the Aggregate
+
+    Parameters
+    ----------
+    observation : Observation
+        The Observation object that is part of the Aggregate
+    effective_from : pandas.Timestamp
+        The effective datetime of when the Observation should be
+        included in the Aggregate
+    effective_until : pandas.Timestamp
+        The effective datetime of when the Observation should be
+        excluded from the Aggregate
+    observation_deleted_at : pandas.Timestamp
+        The datetime that the Observation was deleted from the
+        Arbiter. This indicates that the Observation should be
+        removed from the Aggregate, and without the data
+        from this Observation, the Aggregate is invalid before
+        this time.
+
+    See Also
+    --------
+    Observation
+    Aggregate
+    """
+    observation: Observation
+    effective_from: pd.Timestamp
+    effective_until: Union[pd.Timestamp, None] = None
+    observation_deleted_at: Union[pd.Timestamp, None] = None
+
+    def _special_field_processing(self, model_field, val):
+        if model_field.name in ('effective_until', 'observation_deleted_at'):
+            if val is None:
+                return val
+            else:
+                return pd.Timestamp(val)
+        else:  # pragma: no cover
+            return val
+
+
+def __check_variable__(variable, *args):
+    if not all(arg.variable == variable for arg in args if arg is not None):
```

Comment:
why do we allow `None`?

Reply:
Gets to the issue of when a user can read the aggregate but not the observation. Right now, it's set up to still return an AggregateObservation with the proper times, but observation is None. Better would be to move the NOne check so it's clear that an observation might be None until we have better missing object handling